### PR TITLE
Record allocation spans inside force_allocation

### PR DESCRIPTION
**Pull request by saethlin**
This expands https://github.com/rust-lang/miri/pull/2940 to cover locals

r? @RalfJung

### CODE REVIEW

**Review by RalfJung — COMMENTED** [review]
Thanks, looks great overall with minor nits!

Can you confirm that if you use `current_span` instead, at least one of the tests changes its output?

File: src/tools/miri/tests/fail/dangling_pointers/stack_temporary.stderr [hunk 1]
```diff
@@ -6,7 +6,17 @@ LL |         let val = *x;
...
+help: ALLOC was allocated here:
+  --> $DIR/stack_temporary.rs:LL:CC
+   |
+LL |         let x = make_ref(&mut 0); // The temporary storing "0" is deallocated at the ";"!
+   |                               ^
+help: ALLOC was deallocated here:
+  --> $DIR/stack_temporary.rs:LL:CC
+   |
+LL |         let x = make_ref(&mut 0); // The temporary storing "0" is deallocated at the ";"!
+   |                                 ^
```
[the remark this answers is not in the document]
Comment (saethlin) [reply src/tools/miri/tests/fail/dangling_pointers/stack_temporary.stderr:18]: Yeah! It took me a long time to learn about temporary lifetimes, and it's so neat that we can just point them out so clearly now.

**Comment by saethlin** [comment]
> Can you confirm that if you use `current_span` instead, at least one of the tests changes its output?

Their output does change, but not in a way that is meaningful. I'm adding a more direct test for this.

**Comment by RalfJung** [comment]
Looks good, thanks!
@bors r+

**Pull request merged**